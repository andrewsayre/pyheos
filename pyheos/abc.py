"""Define abstract base classes for HEOS."""

from abc import ABC
from typing import Any


class RemoveHeosFieldABC(ABC):
    """Define an abstract base class that removes the 'heos' from dataclass's fields list to prevent serialization."""

    def __post_init__(self, *args: Any, **kwargs: Any) -> None:
        """Post initialize the player."""
        # Prevent the heos instance from being serialized
        fields = self.__dataclass_fields__.copy()  # type: ignore[has-type] # pylint: disable=access-member-before-definition
        del fields["heos"]
        self.__dataclass_fields__ = fields
