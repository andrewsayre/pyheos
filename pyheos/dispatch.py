"""Defines the dispatch component for notifying others of signals."""

import asyncio
from collections import defaultdict
import functools
from typing import Any, Callable, Dict, List, Sequence

TargetType = Callable[..., Any]
DisconnectType = Callable[[], None]
ConnectType = Callable[[str, TargetType], DisconnectType]
SendType = Callable[..., Sequence[asyncio.Future]]


class Dispatcher:
    """Define the dispatch class."""

    def __init__(self, *, connect: ConnectType = None, send: SendType = None,
                 signal_prefix: str = '', loop=None):
        """Create a new instance of the dispatch component."""
        self._signal_prefix = signal_prefix
        self._signals = defaultdict(list)
        self._loop = loop or asyncio.get_event_loop()
        self._connect = connect or self._default_connect
        self._send = send or self._default_send
        self._disconnects = []

    def connect(self, signal: str, target: TargetType) \
            -> DisconnectType:
        """Connect function to signal.  Must be ran in the event loop."""
        disconnect = self._connect(self._signal_prefix + signal, target)
        self._disconnects.append(disconnect)
        return disconnect

    def send(self, signal: str, *args: Any) -> Sequence[asyncio.Future]:
        """Fire a signal.  Must be ran in the event loop."""
        return self._send(
            self._signal_prefix + signal, *args)

    def disconnect_all(self):
        """Disconnect all connected."""
        disconnects = self._disconnects.copy()
        self._disconnects.clear()
        for disconnect in disconnects:
            disconnect()

    def _default_connect(self, signal: str, target: TargetType) \
            -> DisconnectType:
        """Connect function to signal.  Must be ran in the event loop."""
        self._signals[signal].append(target)

        def remove_dispatcher() -> None:
            """Remove signal listener."""
            try:
                self._signals[signal].remove(target)
            except ValueError:
                # signal was already removed
                pass
        return remove_dispatcher

    def _default_send(self, signal: str, *args: Any) -> \
            Sequence[asyncio.Future]:
        """Fire a signal.  Must be ran in the event loop."""
        targets = self._signals[signal]
        futures = []
        for target in targets:
            task = self._call_target(target, *args)
            futures.append(task)
        return futures

    def _call_target(self, target, *args) -> asyncio.Future:
        check_target = target
        while isinstance(check_target, functools.partial):
            check_target = check_target.func
        if asyncio.iscoroutinefunction(check_target):
            return self._loop.create_task(target(*args))
        return self._loop.run_in_executor(None, target, *args)

    @property
    def signals(self) -> Dict[str, List[TargetType]]:
        """Get the dictionary of registered signals and callbacks."""
        return self._signals
