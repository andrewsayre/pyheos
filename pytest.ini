[pytest]
log_level=DEBUG
testpaths = tests
norecursedirs = .git
timeout = 60
asyncio_default_fixture_loop_scope = function