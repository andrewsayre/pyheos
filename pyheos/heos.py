"""Define the heos manager module."""
import asyncio
import logging
from typing import Dict, Optional, Sequence

from . import const
from .connection import HeosConnection
from .dispatch import Dispatcher
from .player import HeosPlayer
from .response import HeosResponse
from .source import HeosSource, InputSource

_LOGGER = logging.getLogger(__name__)


class Heos:
    """The Heos class provides access to the CLI API."""

    def __init__(self, host: str, *,
                 timeout: float = const.DEFAULT_TIMEOUT,
                 heart_beat: Optional[float] = const.DEFAULT_HEART_BEAT,
                 all_progress_events=True,
                 dispatcher: Dispatcher = None):
        """Init a new instance of the Heos CLI API."""
        self._connection = HeosConnection(
            self, host, timeout=timeout, heart_beat=heart_beat,
            all_progress_events=all_progress_events)
        self._dispatcher = dispatcher or Dispatcher()
        self._players = {}  # type: Dict[int, HeosPlayer]
        self._players_loaded = False
        self._music_sources = {}  # type: Dict[int, HeosSource]
        self._music_sources_loaded = False

    async def connect(self, *, auto_reconnect=False,
                      reconnect_delay: float = const.DEFAULT_RECONNECT_DELAY):
        """Connect to the CLI."""
        await self._connection.connect(auto_reconnect=auto_reconnect,
                                       reconnect_delay=reconnect_delay)

    async def disconnect(self):
        """Disconnect from the CLI."""
        await self._connection.disconnect()

    async def _handle_event(self, event: HeosResponse) -> bool:
        """Handle a heos event."""
        if event.command == const.EVENT_PLAYERS_CHANGED \
                and self._players_loaded:
            await self.get_players(refresh=True)
        if event.command == const.EVENT_SOURCES_CHANGED \
                and self._music_sources_loaded:
            await self.get_music_sources(refresh=True)
        return True

    async def get_players(self, *, refresh=False) -> Dict[int, HeosPlayer]:
        """Get available players."""
        # get players and pull initial state
        if not self._players or refresh:
            payload = await self._connection.commands.get_players()
            players = {}
            player_data = {}
            for data in payload:
                player = HeosPlayer(self, data)
                players[player.player_id] = player
                player_data[player.player_id] = data
            # Match to existing
            for player_id, player in self._players.items():
                if player_id in players:
                    players.pop(player_id)
                    player.set_available(True)
                    player.from_data(player_data[player_id])
                else:
                    player.set_available(False)

            self._players.update(players)
            # Update all statuses
            await asyncio.gather(*[player.refresh() for player in
                                   self._players.values() if player.available])
            self._players_loaded = True
        return self._players

    async def get_music_sources(self, refresh=True) -> Dict[int, HeosSource]:
        """Get available music sources."""
        if not self._music_sources or refresh:
            payload = await self._connection.commands.get_music_sources()
            self._music_sources.clear()
            for data in payload:
                source = HeosSource(self._connection.commands, data)
                self._music_sources[source.source_id] = source
            self._music_sources_loaded = True
        return self._music_sources

    async def get_input_sources(self) -> Sequence[InputSource]:
        """Get available input sources."""
        payload = await self._connection.commands.browse(
            const.MUSIC_SOURCE_AUX_INPUT)
        sources = [HeosSource(self._connection.commands, item)
                   for item in payload]
        input_sources = []
        for source in sources:
            player_id = source.source_id
            items = await source.browse()
            input_sources.extend(
                [InputSource(player_id, item.name, item.media_id)
                 for item in items])
        return input_sources

    async def get_favorites(self) -> Dict[int, HeosSource]:
        """Get available favorites."""
        payload = await self._connection.commands.browse(
            const.MUSIC_SOURCE_FAVORITES)
        sources = [HeosSource(self._connection.commands, item)
                   for item in payload]
        return {index + 1: source for index, source in enumerate(sources)}

    @property
    def dispatcher(self) -> Dispatcher:
        """Get the dispatcher instance."""
        return self._dispatcher

    @property
    def players(self) -> Dict[int, HeosPlayer]:
        """Get the available players."""
        return self._players

    @property
    def music_sources(self) -> Dict[int, HeosSource]:
        """Get available music sources."""
        return self._music_sources

    @property
    def connection_state(self):
        """Get the state of the connection."""
        return self._connection.state
