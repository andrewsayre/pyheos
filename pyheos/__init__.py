"""pyheos - a library for interacting with HEOS devices."""
from . import const
from .dispatch import Dispatcher
from .heos import Heos
from .player import HeosNowPlayingMedia, HeosPlayer
from .response import CommandError
from .source import HeosSource, InputSource

__all__ = [
    'const',
    'CommandError',
    'Dispatcher',
    'Heos',
    'HeosPlayer',
    'HeosNowPlayingMedia',
    'HeosSource',
    'InputSource'
]
