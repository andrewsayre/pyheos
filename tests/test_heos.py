"""Tests for the heos class."""
import asyncio

import pytest

from pyheos import const
from pyheos.dispatch import Dispatcher
from pyheos.heos import Heos

from . import connect_handler, get_fixture


def test_init():
    """Test init sets properties."""
    heos = Heos('127.0.0.1')
    assert isinstance(heos.dispatcher, Dispatcher)
    assert heos.players == {}
    assert heos.music_sources == {}
    assert heos.connection_state == const.STATE_DISCONNECTED


@pytest.mark.asyncio
async def test_connect(mock_device):
    """Test connect updates state and fires signal."""
    heos = Heos('127.0.0.1')
    signal = connect_handler(heos, const.SIGNAL_HEOS_EVENT,
                             const.EVENT_CONNECTED)
    await heos.connect()
    await signal.wait()
    # Assert 1st connection is used for commands
    assert heos.connection_state == const.STATE_CONNECTED
    assert len(mock_device.connections) == 1
    connection = mock_device.connections[0]
    assert connection.is_registered_for_events

    await heos.disconnect()


@pytest.mark.asyncio
async def test_heart_beat(mock_device):
    """Test heart beat fires at interval."""
    heos = Heos('127.0.0.1', heart_beat=0.5)
    await heos.connect()
    await asyncio.sleep(1.0)

    connection = mock_device.connections[0]
    assert len(connection.commands[const.COMMAND_HEART_BEAT]) == 1

    await heos.disconnect()


@pytest.mark.asyncio
async def test_connect_fails():
    """Test connect fails when host not available."""
    heos = Heos('127.0.0.1')
    with pytest.raises(ConnectionRefusedError):
        await heos.connect()
    # Also fails for initial connection even with reconnect.
    with pytest.raises(ConnectionRefusedError):
        await heos.connect(auto_reconnect=True)


@pytest.mark.asyncio
async def test_connect_timeout():
    """Test connect fails when host not available."""
    heos = Heos('www.google.com', timeout=1)
    with pytest.raises(asyncio.TimeoutError):
        await heos.connect()
    # Also fails for initial connection even with reconnect.
    with pytest.raises(asyncio.TimeoutError):
        await heos.connect(auto_reconnect=True)


@pytest.mark.asyncio
async def test_disconnect(mock_device, heos):
    """Test disconnect updates state and fires signal."""
    # Fixture automatically connects
    signal = connect_handler(heos, const.SIGNAL_HEOS_EVENT,
                             const.EVENT_DISCONNECTED)
    await heos.disconnect()
    await signal.wait()
    assert heos.connection_state == const.STATE_DISCONNECTED


@pytest.mark.asyncio
async def test_connection_error_during_event(mock_device, heos):
    """Test connection error during event results in disconnected."""
    disconnect_signal = connect_handler(
        heos, const.SIGNAL_HEOS_EVENT, const.EVENT_DISCONNECTED)

    # Assert transitions to reconnecting and fires disconnect
    await mock_device.stop()
    await disconnect_signal.wait()
    assert heos.connection_state == const.STATE_DISCONNECTED


@pytest.mark.asyncio
async def test_connection_error_during_command(mock_device, heos):
    """Test connection error during command results in disconnected."""
    disconnect_signal = connect_handler(
        heos, const.SIGNAL_HEOS_EVENT, const.EVENT_DISCONNECTED)

    # Assert transitions to reconnecting and fires disconnect
    await mock_device.stop()
    with pytest.raises(asyncio.TimeoutError):
        await heos.get_players()

    await disconnect_signal.wait()
    assert heos.connection_state == const.STATE_DISCONNECTED


@pytest.mark.asyncio
async def test_reconnect_during_event(mock_device):
    """Test reconnect while waiting for events/responses."""
    heos = Heos('127.0.0.1', timeout=0.1)

    connect_signal = connect_handler(
        heos, const.SIGNAL_HEOS_EVENT, const.EVENT_CONNECTED)
    disconnect_signal = connect_handler(
        heos, const.SIGNAL_HEOS_EVENT, const.EVENT_DISCONNECTED)

    # Assert open and fires connected
    await heos.connect(auto_reconnect=True, reconnect_delay=0)
    await connect_signal.wait()
    assert heos.connection_state == const.STATE_CONNECTED
    connect_signal.clear()

    # Assert transitions to reconnecting and fires disconnect
    await mock_device.stop()
    await disconnect_signal.wait()
    assert heos.connection_state == const.STATE_RECONNECTING

    # Assert reconnects once server is back up and fires connected
    await asyncio.sleep(0.5)  # Force reconnect timeout
    await mock_device.start()
    await connect_signal.wait()
    assert heos.connection_state == const.STATE_CONNECTED

    await heos.disconnect()


@pytest.mark.asyncio
async def test_reconnect_during_command(mock_device):
    """Test reconnect while waiting for events/responses."""
    heos = Heos('127.0.0.1', timeout=1.0)

    connect_signal = connect_handler(
        heos, const.SIGNAL_HEOS_EVENT, const.EVENT_CONNECTED)
    disconnect_signal = connect_handler(
        heos, const.SIGNAL_HEOS_EVENT, const.EVENT_DISCONNECTED)

    # Assert open and fires connected
    await heos.connect(auto_reconnect=True)
    await connect_signal.wait()
    assert heos.connection_state == const.STATE_CONNECTED
    connect_signal.clear()

    # Act
    await mock_device.stop()
    await mock_device.start()
    with pytest.raises(asyncio.TimeoutError):
        await heos.get_players()

    # Assert signals set
    await disconnect_signal.wait()
    await connect_signal.wait()
    assert heos.connection_state == const.STATE_CONNECTED

    await heos.disconnect()


@pytest.mark.asyncio
async def test_reconnect_cancelled(mock_device):
    """Test reconnect is canceled by calling disconnect."""
    heos = Heos('127.0.0.1')

    connect_signal = connect_handler(
        heos, const.SIGNAL_HEOS_EVENT, const.EVENT_CONNECTED)
    disconnect_signal = connect_handler(
        heos, const.SIGNAL_HEOS_EVENT, const.EVENT_DISCONNECTED)

    # Assert open and fires connected
    await heos.connect(auto_reconnect=True)
    await connect_signal.wait()
    assert heos.connection_state == const.STATE_CONNECTED
    connect_signal.clear()

    # Assert transitions to reconnecting and fires disconnect
    await mock_device.stop()
    await disconnect_signal.wait()
    assert heos.connection_state == const.STATE_RECONNECTING

    # Assert reconnects once server is back up and fires connected
    await heos.disconnect()
    assert heos.connection_state == const.STATE_DISCONNECTED


@pytest.mark.asyncio
async def test_get_players(mock_device, heos):
    """Test the get_players method load players."""
    await heos.get_players()
    # Assert players loaded
    assert len(heos.players) == 2
    player = heos.players.get(1)
    assert player.player_id == 1
    assert player.name == 'Back Patio'
    assert player.ip_address == '192.168.0.1'
    assert player.line_out == 1
    assert player.model == 'HEOS Drive'
    assert player.network == 'wired'
    assert player.state == const.PLAY_STATE_STOP
    assert player.version == '1.493.180'
    assert player.volume == 36
    assert not player.is_muted
    assert player.repeat == const.REPEAT_OFF
    assert not player.shuffle
    assert player.available
    assert player.heos == heos


@pytest.mark.asyncio
async def test_player_state_changed_event(mock_device, heos):
    """Test playing state updates when event is received."""
    # assert not playing
    await heos.get_players()
    player = heos.players.get(1)
    assert player.state == const.PLAY_STATE_STOP

    # Attach dispatch handler
    signal = asyncio.Event()

    async def handler(player_id: int, event: str):
        assert player_id == player.player_id
        assert event == const.EVENT_PLAYER_STATE_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_PLAYER_EVENT, handler)

    # Write event through mock device
    event_to_raise = (await get_fixture("event.player_state_changed")) \
        .replace("{player_id}", str(player.player_id)) \
        .replace("{state}", const.PLAY_STATE_PLAY)
    await mock_device.write_event(event_to_raise)

    # Wait until the signal
    await signal.wait()
    # Assert state changed
    assert player.state == const.PLAY_STATE_PLAY
    assert heos.players.get(2).state == const.PLAY_STATE_STOP


@pytest.mark.asyncio
async def test_player_now_playing_changed_event(mock_device, heos):
    """Test now playing updates when event is received."""
    # assert current state
    await heos.get_players()
    player = heos.players.get(1)
    now_playing = player.now_playing_media
    assert now_playing.album == ''
    assert now_playing.type == 'song'
    assert now_playing.album_id == ''
    assert now_playing.artist == ''
    assert now_playing.image_url == ''
    assert now_playing.media_id == 'catalog/playlists/genres'
    assert now_playing.queue_id == 1
    assert now_playing.source_id == 13
    assert now_playing.song == 'Disney Hits'
    assert now_playing.station is None
    assert now_playing.supported_controls == const.CONTROLS_ALL

    # Attach dispatch handler
    signal = asyncio.Event()

    async def handler(player_id: int, event: str):
        assert player_id == player.player_id
        assert event == const.EVENT_PLAYER_NOW_PLAYING_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_PLAYER_EVENT, handler)

    # Write event through mock device
    mock_device.register(const.COMMAND_GET_NOW_PLAYING_MEDIA, None,
                         "player.get_now_playing_media_changed",
                         replace=True)
    event_to_raise = (await get_fixture("event.player_now_playing_changed")) \
        .replace("{player_id}", str(player.player_id))
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()
    # Assert state changed
    assert now_playing.album == "I've Been Waiting (Single) (Explicit)"
    assert now_playing.type == 'station'
    assert now_playing.album_id == '1'
    assert now_playing.artist == 'Lil Peep & ILoveMakonnen'
    assert now_playing.image_url == 'http://media/url'
    assert now_playing.media_id == '2PxuY99Qty'
    assert now_playing.queue_id == 1
    assert now_playing.source_id == 1
    assert now_playing.song == "I've Been Waiting (feat. Fall Out Boy)"
    assert now_playing.station == "Today's Hits Radio"
    assert now_playing.current_position is None
    assert now_playing.current_position_updated is None
    assert now_playing.duration is None
    assert now_playing.supported_controls == const.CONTROLS_FORWARD_ONLY


@pytest.mark.asyncio
async def test_player_volume_changed_event(mock_device, heos):
    """Test volume state updates when event is received."""
    # assert not playing
    await heos.get_players()
    player = heos.players.get(1)
    assert player.volume == 36
    assert not player.is_muted

    # Attach dispatch handler
    signal = asyncio.Event()

    async def handler(player_id: int, event: str):
        assert player_id == player.player_id
        assert event == const.EVENT_PLAYER_VOLUME_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_PLAYER_EVENT, handler)

    # Write event through mock device
    event_to_raise = (await get_fixture("event.player_volume_changed")) \
        .replace("{player_id}", str(player.player_id)) \
        .replace("{level}", '50') \
        .replace("{mute}", 'on')
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()
    # Assert state changed
    assert player.volume == 50
    assert player.is_muted
    assert heos.players.get(2).volume == 36
    assert not heos.players.get(2).is_muted


@pytest.mark.asyncio
async def test_player_now_playing_progress_event(mock_device, heos):
    """Test now playing progress updates when event received."""
    # assert not playing
    await heos.get_players()
    player = heos.players.get(1)
    assert player.now_playing_media.duration is None
    assert player.now_playing_media.current_position is None
    assert player.now_playing_media.current_position_updated is None

    # Attach dispatch handler
    signal = asyncio.Event()

    async def handler(player_id: int, event: str):
        assert player_id == player.player_id
        assert event == const.EVENT_PLAYER_NOW_PLAYING_PROGRESS
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_PLAYER_EVENT, handler)

    # Write event through mock device
    event_to_raise = (await get_fixture("event.player_now_playing_progress")) \
        .replace("{player_id}", str(player.player_id)) \
        .replace("{cur_pos}", '113000') \
        .replace("{duration}", '210000')
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set or timeout
    await signal.wait()
    # Assert state changed
    assert player.now_playing_media.duration == 210000
    assert player.now_playing_media.current_position == 113000
    assert player.now_playing_media.current_position_updated is not None
    player2 = heos.players.get(2)
    assert player2.now_playing_media.duration is None
    assert player2.now_playing_media.current_position is None
    assert player2.now_playing_media.current_position_updated is None


@pytest.mark.asyncio
async def test_limited_progress_event_updates(mock_device):
    """Test progress updates only once if no other events."""
    # assert not playing
    heos = Heos('127.0.0.1', all_progress_events=False)
    await heos.connect()
    await heos.get_players()
    player = heos.players.get(1)

    # Attach dispatch handler
    signal = asyncio.Event()

    async def handler(player_id: int, event: str):
        if not signal.is_set():
            signal.set()
        else:
            pytest.fail("Handler invoked more than once.")
    heos.dispatcher.connect(const.SIGNAL_PLAYER_EVENT, handler)

    # Write event through mock device
    event_to_raise = (await get_fixture("event.player_now_playing_progress")) \
        .replace("{player_id}", str(player.player_id)) \
        .replace("{cur_pos}", '113000') \
        .replace("{duration}", '210000')

    # raise it multiple times.
    await mock_device.write_event(event_to_raise)
    await mock_device.write_event(event_to_raise)
    await mock_device.write_event(event_to_raise)
    await signal.wait()


@pytest.mark.asyncio
async def test_repeat_mode_changed_event(mock_device, heos):
    """Test repeat mode changes when event received."""
    # assert not playing
    await heos.get_players()
    player = heos.players.get(1)
    assert player.repeat == const.REPEAT_OFF

    # Attach dispatch handler
    signal = asyncio.Event()

    async def handler(player_id: int, event: str):
        assert player_id == player.player_id
        assert event == const.EVENT_REPEAT_MODE_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_PLAYER_EVENT, handler)

    # Write event through mock device
    event_to_raise = await get_fixture("event.repeat_mode_changed")
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()
    # Assert state changed
    assert player.repeat == const.REPEAT_ON_ALL


@pytest.mark.asyncio
async def test_shuffle_mode_changed_event(mock_device, heos):
    """Test shuffle mode changes when event received."""
    # assert not playing
    await heos.get_players()
    player = heos.players.get(1)
    assert not player.shuffle

    # Attach dispatch handler
    signal = asyncio.Event()

    async def handler(player_id: int, event: str):
        assert player_id == player.player_id
        assert event == const.EVENT_SHUFFLE_MODE_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_PLAYER_EVENT, handler)

    # Write event through mock device
    event_to_raise = await get_fixture("event.shuffle_mode_changed")
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()
    # Assert state changed
    assert player.shuffle


@pytest.mark.asyncio
async def test_players_changed_event(mock_device, heos):
    """Test players are resynced when event received."""
    # assert not playing
    old_players = (await heos.get_players()).copy()

    # Attach dispatch handler
    signal = asyncio.Event()

    async def handler(event: str):
        assert event == const.EVENT_PLAYERS_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_CONTROLLER_EVENT, handler)

    # Write event through mock device
    mock_device.register(const.COMMAND_GET_PLAYERS, None,
                         'player.get_players_changed', replace=True)
    event_to_raise = await get_fixture("event.players_changed")
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()
    assert len(heos.players) == 3
    # Assert 2 (Front Porch) was marked unavailable
    assert not old_players[2].available
    assert old_players[2] == heos.players[2]
    # Assert 3 (Basement) was added
    assert heos.players.get(3) is not None
    # Assert 1 (Backyard) was updated
    assert heos.players.get(1).name == 'Backyard'


@pytest.mark.asyncio
async def test_sources_changed_event(mock_device, heos):
    """Test sources changed fires dispatcher."""
    mock_device.register(const.COMMAND_BROWSE_GET_SOURCES, None,
                         'browse.get_music_sources')
    await heos.get_music_sources()
    signal = asyncio.Event()

    async def handler(event: str):
        assert event == const.EVENT_SOURCES_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_CONTROLLER_EVENT, handler)

    # Write event through mock device
    mock_device.register(const.COMMAND_BROWSE_GET_SOURCES, None,
                         'browse.get_music_sources_changed', replace=True)
    event_to_raise = await get_fixture("event.sources_changed")
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()
    assert heos.music_sources[const.MUSIC_SOURCE_TUNEIN].available


@pytest.mark.asyncio
async def test_groups_changed_event(mock_device, heos):
    """Test groups changed fires dispatcher."""
    signal = asyncio.Event()

    async def handler(event: str):
        assert event == const.EVENT_GROUPS_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_CONTROLLER_EVENT, handler)

    # Write event through mock device
    event_to_raise = await get_fixture("event.groups_changed")
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()


@pytest.mark.asyncio
async def test_player_playback_error_event(mock_device, heos):
    """Test player playback error fires dispatcher."""
    await heos.get_players()
    signal = asyncio.Event()

    async def handler(player_id: int, event: str):
        assert player_id == 1
        assert event == const.EVENT_PLAYER_PLAYBACK_ERROR
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_PLAYER_EVENT, handler)

    # Write event through mock device
    event_to_raise = await get_fixture("event.player_playback_error")
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()
    assert heos.players[1].playback_error == 'Could Not Download'


@pytest.mark.asyncio
async def test_player_queue_changed_event(mock_device, heos):
    """Test player queue changed fires dispatcher."""
    await heos.get_players()
    signal = asyncio.Event()

    async def handler(player_id: int, event: str):
        assert player_id == 1
        assert event == const.EVENT_PLAYER_QUEUE_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_PLAYER_EVENT, handler)

    # Write event through mock device
    event_to_raise = await get_fixture("event.player_queue_changed")
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()


@pytest.mark.asyncio
async def test_group_volume_changed_event(mock_device, heos):
    """Test group volume changed fires dispatcher."""
    signal = asyncio.Event()

    async def handler(group_id: int, event: str):
        assert group_id == 1
        assert event == const.EVENT_GROUP_VOLUME_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_GROUP_EVENT, handler)

    # Write event through mock device
    event_to_raise = await get_fixture("event.group_volume_changed")
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()


@pytest.mark.asyncio
async def test_user_changed_event(mock_device, heos):
    """Test user changed fires dispatcher."""
    signal = asyncio.Event()

    async def handler(event: str):
        assert event == const.EVENT_USER_CHANGED
        signal.set()
    heos.dispatcher.connect(const.SIGNAL_CONTROLLER_EVENT, handler)

    # Write event through mock device
    event_to_raise = await get_fixture("event.user_changed")
    await mock_device.write_event(event_to_raise)

    # Wait until the signal is set
    await signal.wait()


@pytest.mark.asyncio
async def test_get_music_sources(mock_device, heos):
    """Test the heos connect method."""
    mock_device.register(const.COMMAND_BROWSE_GET_SOURCES, None,
                         'browse.get_music_sources')

    sources = await heos.get_music_sources()
    assert len(sources) == 15
    pandora = sources[const.MUSIC_SOURCE_PANDORA]
    assert pandora.source_id == 1
    assert pandora.image_url == 'https://production.ws.skyegloup.com:443/' \
                                'media/images/service/logos/pandora.png'
    assert pandora.type == const.TYPE_MUSIC_SERVICE
    assert pandora.available
    assert pandora.service_username == 'test@test.com'
    assert not pandora.container
    assert not pandora.playable


@pytest.mark.asyncio
async def test_get_input_sources(mock_device, heos):
    """Test the get input sources method."""
    mock_device.register(const.COMMAND_BROWSE_BROWSE, {'sid': '1027'},
                         'browse.browse_aux_input')
    mock_device.register(const.COMMAND_BROWSE_BROWSE, {'sid': '546978854'},
                         'browse.browse_theater_receiver')
    mock_device.register(const.COMMAND_BROWSE_BROWSE, {'sid': '-263109739'},
                         'browse.browse_heos_drive')

    sources = await heos.get_input_sources()
    assert len(sources) == 18
    source = sources[0]
    assert source.name == 'Theater Receiver - CBL/SAT'
    assert source.input_name == const.INPUT_CABLE_SAT
    assert source.player_id == 546978854


@pytest.mark.asyncio
async def test_get_favorites(mock_device, heos):
    """Test the get favorites method."""
    mock_device.register(const.COMMAND_BROWSE_BROWSE, {'sid': '1028'},
                         'browse.browse_favorites')

    sources = await heos.get_favorites()
    assert len(sources) == 3
    assert sorted(sources.keys()) == [1, 2, 3]
    fav = sources[1]
    assert fav.playable
    assert fav.name == 'Thumbprint Radio'
    assert fav.type == const.TYPE_STATION
