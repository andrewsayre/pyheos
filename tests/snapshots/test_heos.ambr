# serializer version: 1
# name: test_get_players
  dict({
    1: dict({
      'available': True,
      'control': <VolumeControlType.IR: 2>,
      'group_id': 2,
      'ip_address': '127.0.0.1',
      'is_muted': False,
      'line_out': <LineOutLevelType.FIXED: 2>,
      'model': 'HEOS Drive',
      'name': 'Back Patio',
      'network': <NetworkType.WIRED: 'wired'>,
      'now_playing_media': dict({
        'album': 'Album',
        'album_id': '123456',
        'artist': 'Artist',
        'current_position': None,
        'current_position_updated': None,
        'duration': None,
        'image_url': 'http://cont-5.p-cdn.us/images/public/int/6/1/1/9/050087149116_500W_500H.jpg',
        'media_id': '4256592506324148495',
        'options': list([
          dict({
            'context': 'play',
            'id': 11,
            'name': 'Thumbs Up',
          }),
          dict({
            'context': 'play',
            'id': 12,
            'name': 'Thumbs Down',
          }),
          dict({
            'context': 'play',
            'id': 19,
            'name': 'Add to HEOS Favorites',
          }),
        ]),
        'queue_id': 1,
        'song': "Disney (Children's) Radio",
        'source_id': 13,
        'station': "Disney (Children's) Radio",
        'supported_controls': list([
          <ControlType.PLAY: 'play'>,
          <ControlType.PAUSE: 'pause'>,
          <ControlType.STOP: 'stop'>,
          <ControlType.PLAY_NEXT: 'play_next'>,
          <ControlType.PLAY_PREVIOUS: 'play_previous'>,
        ]),
        'type': 'station',
      }),
      'playback_error': None,
      'player_id': 1,
      'repeat': <RepeatType.OFF: 'off'>,
      'serial': 'B1A2C3K',
      'shuffle': False,
      'state': <PlayState.STOP: 'stop'>,
      'version': '1.493.180',
      'volume': 36,
    }),
    2: dict({
      'available': True,
      'control': <VolumeControlType.UNKNOWN: 0>,
      'group_id': 2,
      'ip_address': '127.0.0.2',
      'is_muted': False,
      'line_out': <LineOutLevelType.VARIABLE: 1>,
      'model': 'HEOS Drive',
      'name': 'Front Porch',
      'network': <NetworkType.WIFI: 'wifi'>,
      'now_playing_media': dict({
        'album': 'Album',
        'album_id': '123456',
        'artist': 'Artist',
        'current_position': None,
        'current_position_updated': None,
        'duration': None,
        'image_url': 'http://cont-5.p-cdn.us/images/public/int/6/1/1/9/050087149116_500W_500H.jpg',
        'media_id': '4256592506324148495',
        'options': list([
          dict({
            'context': 'play',
            'id': 11,
            'name': 'Thumbs Up',
          }),
          dict({
            'context': 'play',
            'id': 12,
            'name': 'Thumbs Down',
          }),
          dict({
            'context': 'play',
            'id': 19,
            'name': 'Add to HEOS Favorites',
          }),
        ]),
        'queue_id': 1,
        'song': "Disney (Children's) Radio",
        'source_id': 13,
        'station': "Disney (Children's) Radio",
        'supported_controls': list([
          <ControlType.PLAY: 'play'>,
          <ControlType.PAUSE: 'pause'>,
          <ControlType.STOP: 'stop'>,
          <ControlType.PLAY_NEXT: 'play_next'>,
          <ControlType.PLAY_PREVIOUS: 'play_previous'>,
        ]),
        'type': 'station',
      }),
      'playback_error': None,
      'player_id': 2,
      'repeat': <RepeatType.OFF: 'off'>,
      'serial': None,
      'shuffle': False,
      'state': <PlayState.STOP: 'stop'>,
      'version': '1.493.180',
      'volume': 36,
    }),
  })
# ---
